ledger/complete/mtrie/trie/trie.go

```diff
@@ -153,135 +164,98 @@ func (mt *MTrie) read(head *node.Node, paths []ledger.Path) []*ledger.Payload {
 //   * subtries that remain unchanged are from the parent trie instead of copied.
 // UNSAFE: method requires the following conditions to be satisfied:
 //   * keys are NOT duplicated
-// TODO: move consistency checks from MForest to here, to make API is safe and self-contained
+// TODO: move consistency checks from MForest to here, to make API safe and self-contained
 func NewTrieWithUpdatedRegisters(parentTrie *MTrie, updatedPaths []ledger.Path, updatedPayloads []ledger.Payload) (*MTrie, error) {
 	parentRoot := parentTrie.root
-	updatedRoot, err := update(parentTrie.height, parentRoot.Height(), parentRoot, updatedPaths, updatedPayloads)
-	if err != nil {
-		return nil, fmt.Errorf("constructing updated trie failed: %w", err)
-	}
+	updatedRoot := parentTrie.update(parentRoot.Height(), parentRoot, updatedPaths, updatedPayloads, nil)
 	updatedTrie, err := NewMTrie(updatedRoot)
 	if err != nil {
 		return nil, fmt.Errorf("constructing updated trie failed: %w", err)
 	}
 	return updatedTrie, nil
 }
 
-// update returns the head of updated sub-trie for the specified key-value pairs.
-// UNSAFE: update requires the following conditions to be satisfied,
-// but does not explicitly check them for performance reasons
-//   * all keys AND the parent node share the same common prefix [0 : mt.maxHeight-1 - headHeight)
-//     (excluding the bit at index headHeight)
-//   * keys are NOT duplicated
-// TODO: remove error return
-func update(treeHeight int, nodeHeight int, parentNode *node.Node, paths []ledger.Path, payloads []ledger.Payload) (*node.Node, error) {
-	if parentNode == nil { // parent Trie has no sub-trie for the set of paths => construct entire subtree
-		return constructSubtrie(treeHeight, nodeHeight, paths, payloads)
+func (parentTrie *MTrie) update(nodeHeight int, parentNode *node.Node,
+	paths []ledger.Path, payloads []ledger.Payload, compactLeaf *node.Node) *node.Node {
+
+	// No new paths to write
+	if len(paths) == 0 {
+		// check is a compactLeaf from a higher height is still left
+		if compactLeaf != nil {
+			return node.NewLeaf(compactLeaf.Path(), compactLeaf.Payload(), nodeHeight)
+		}
+		return parentNode
 	}
 
-	if len(paths) == 0 { // We are not changing any values in this sub-trie => return parent trie
-		return parentNode, nil
+	if len(paths) == 1 && parentNode == nil && compactLeaf == nil {
```

Comment:
I'm concerned about these `if` conditions, looks a bit complex. Is there way to simplify?

Reply:
Unfortunately, there are many stop conditions for this recursion. And to create a new leaf, the 3 conditions need to be met, which I agree looks complex. 

Btw, I classified the 3 conditions from less likely to more likely so that evaluating the boolean expression stops early (if `len(paths)`>1` we don't even evaluate the rest)

Reply:
can we handle the `parentNode == nil` first:

```
if parentNode == nil {
  if len(paths) == 1 && compactLeaf == nil {
    return NewLeaf(...)
  }
  return ... // what do we return in this case?
}

// now parentNode != nil, and we don't need to check it anymore
```

Reply:
The issue here is that this code `return ... // what do we return in this case?` would be to the main core of the recursion (split paths, recurse on left and right in parallel, create a new node..),  which is then needed again when parentNode != nil